plat/imx/common/imx_sip_handler.c

```diff
@@ -188,6 +196,49 @@ int imx_src_handler(uint32_t smc_fid,
 	case IMX_SIP_SRC_IS_M_CORE_STARTED:
 		val = mmio_read_32(IMX_IOMUX_GPR_BASE + IOMUXC_GPR22);
 		return !(val & GPR_M7_CPUWAIT);
+	case IMX_SIP_SRC_STOP_M_CORE:
+		/*
+		 * Safe stop
+		 *    If M4 already in WFI,  perform below steps.
+		 * a)	Set [0x303A_002C].0=0   [ request SLEEPHOLDREQn ]
+		 * b)	Wait  [0x303A_00EC].1 = 0  [ wait SLEEPHOLDACKn ]
+		 * c)	Set  GPR.CPUWAIT=1
+		 * d)	Set [0x303A_002C].0=1  [ de-assert SLEEPHOLDREQn ]
+		 * e)	Set SRC_M7_RCR[3:0] = 0xE0   [ reset M7 core/plat ]
+		 * f)	Wait SRC_M7_RCR[3:0] = 0x8
+		 * The following steps move to start part.
+		 * g/h is actually no needed here.
+		 * g)	Init TCM or DDR
+		 * h)	Set GPR.INITVTOR
+		 * i)	Set GPR.CPUWAIT=0,  M7 starting running
+		 */
+		offset = LPS_CPU1;
+		val = mmio_read_32(IMX_GPC_BASE + offset);
+		/* Not in stop/wait mode */
+		if (!(val & (0x3 << 24))) {
+			mmio_clrbits_32(IMX_GPC_BASE + GPC_MISC, M7_SLEEP_HOLD_REQ_B);
+
+			timeout = timeout_init_us(10000);
+			while ((mmio_read_32(IMX_GPC_BASE + offset) & 0x2)) {
+				if (timeout_elapsed(timeout)) {
+					ret1 = -ETIMEDOUT;
+					break;
+				}
+			}
+		}
+		mmio_setbits_32(IMX_IOMUX_GPR_BASE + IOMUXC_GPR22, GPR_M7_CPUWAIT);
+		mmio_setbits_32(IMX_GPC_BASE + GPC_MISC, M7_SLEEP_HOLD_REQ_B);
+		mmio_setbits_32(IMX_SRC_BASE + SRC_M7RCR_OFFSET, 0xE);
+		timeout = timeout_init_us(10000);
+		while ((mmio_read_32(IMX_SRC_BASE + SRC_M7RCR_OFFSET) & 0xF) != 8) {
+			if (timeout_elapsed(timeout)) {
+				ret2 = -ETIMEDOUT;
+				break;
+			}
+		}
+		SMC_SET_GP(handle, CTX_GPREG_X1, ret1);
```

Comment:
Same as above, can be simplified:
```suggestion
		timeout = timeout_init_us(10000);
		ret = 0;
		while ((mmio_read_32(IMX_SRC_BASE + SRC_M7RCR_OFFSET) & 0xF) != 8) {
			if (timeout_elapsed(timeout)) {
				ret2 = -ETIMEDOUT;
				break;
			}
		}
		SMC_SET_GP(handle, CTX_GPREG_X2, ret);
```